File: core/src/main/java/nl/t64/game/rpg/screens/inventory/BaseTable.java
package nl.t64.game.rpg.screens.inventory;

import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.graphics.g2d.BitmapFont;
import com.badlogic.gdx.graphics.g2d.NinePatch;
import com.badlogic.gdx.scenes.scene2d.ui.Image;
import com.badlogic.gdx.scenes.scene2d.ui.Label;
import com.badlogic.gdx.scenes.scene2d.ui.Skin;
import com.badlogic.gdx.scenes.scene2d.ui.Table;
import com.badlogic.gdx.scenes.scene2d.utils.NinePatchDrawable;
import com.badlogic.gdx.utils.Scaling;
import nl.t64.game.rpg.Utils;


abstract class BaseTable {

    private static final String SPRITE_TOP_BORDER = "sprites/top_border.png";
    private static final String TEXT_FONT = "fonts/spectral.ttf";
    private static final int TEXT_SIZE = 20;
    private static final float LINE_HEIGHT = 26f;
    private static final float PADDING = 20f;
    private static final float PADDING_RIGHT = 10f;

    final Table table;
    private final BitmapFont font;

    BaseTable() {
        this.font = Utils.getResourceManager().getTrueTypeAsset(TEXT_FONT, TEXT_SIZE);
        this.table = new Table(createSkin());
        this.table.defaults().height(LINE_HEIGHT);
        this.table.pad(PADDING).padRight(PADDING_RIGHT);
    }

    abstract void render();

    void createBonusFromInventory(int totalFromInventory) {
        if (totalFromInventory > 0) {
            var label = new Label(String.format("+%s", totalFromInventory), new Label.LabelStyle(font, Color.FOREST));
            table.add(label);
        } else if (totalFromInventory < 0) {
            var label = new Label(String.valueOf(totalFromInventory), new Label.LabelStyle(font, Color.FIREBRICK));
            table.add(label);
        } else {
            table.add("");
        }
    }

    void createPreview(int difference) {
        if (difference > 0) {
            var label = new Label(String.format("+%s", difference), new Label.LabelStyle(font, Color.LIME));
            table.add(label).row();
        } else if (difference < 0) {
            var label = new Label(String.valueOf(difference), new Label.LabelStyle(font, Color.SCARLET));
            table.add(label).row();
        } else {
            table.add("").row();
        }
    }

    void setTopBorder(Table table) {
        var texture = Utils.getResourceManager().getTextureAsset(SPRITE_TOP_BORDER);
        var ninepatch = new NinePatch(texture, 0, 0, 1, 0);
        var drawable = new NinePatchDrawable(ninepatch);
        table.setBackground(drawable);
    }

    Image createImageOf(String id) {
        var textureRegion = Utils.getResourceManager().getAtlasTexture(id.toLowerCase());
        var image = new Image(textureRegion);
        image.setScaling(Scaling.none);
        return image;
    }

    private Skin createSkin() {
        Skin skin = new Skin();
        skin.add("default", new Label.LabelStyle(font, Color.BLACK));
        return skin;
    }

}

File: core/src/nl/t64/game/rpg/screens/inventory/BaseTable.kt
package nl.t64.game.rpg.screens.inventory

import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.graphics.g2d.BitmapFont
import com.badlogic.gdx.scenes.scene2d.Stage
import com.badlogic.gdx.scenes.scene2d.ui.Image
import com.badlogic.gdx.scenes.scene2d.ui.Label
import com.badlogic.gdx.scenes.scene2d.ui.Label.LabelStyle
import com.badlogic.gdx.scenes.scene2d.ui.Skin
import com.badlogic.gdx.scenes.scene2d.ui.Table
import com.badlogic.gdx.utils.Scaling
import nl.t64.game.rpg.Utils.resourceManager
import nl.t64.game.rpg.components.party.HeroItem
import nl.t64.game.rpg.screens.inventory.InventoryUtils.getSelectedHero
import nl.t64.game.rpg.screens.inventory.InventoryUtils.setWindowDeselected
import nl.t64.game.rpg.screens.inventory.InventoryUtils.setWindowSelected
import nl.t64.game.rpg.screens.inventory.itemslot.ItemSlot
import nl.t64.game.rpg.screens.inventory.tooltip.BaseTooltip
import nl.t64.game.rpg.screens.inventory.tooltip.PersonalityTooltip


private const val TEXT_FONT = "fonts/spectral_extra_bold_20.ttf"
private const val TEXT_SIZE = 20
private const val LINE_HEIGHT = 26f
private const val PADDING = 20f
private const val PADDING_RIGHT = 10f

abstract class BaseTable(val tooltip: PersonalityTooltip) : WindowSelector {

    val container: Table = Table()
    val font: BitmapFont = resourceManager.getTrueTypeAsset(TEXT_FONT, TEXT_SIZE)
    val table: Table = Table(createSkin()).apply {
        defaults().height(LINE_HEIGHT)
        pad(PADDING).padRight(PADDING_RIGHT)
    }

    lateinit var selectedHero: HeroItem
    var selectedIndex = 0
    var hasJustUpdated = true

    override fun setKeyboardFocus(stage: Stage) {
        selectedHero = getSelectedHero()
        stage.keyboardFocus = table
        setWindowSelected(container)
    }

    override fun getCurrentSlot(): ItemSlot {
        throw IllegalStateException("getCurrentSlot() is not used in BaseTables.")
    }

    override fun getCurrentTooltip(): BaseTooltip {
        return tooltip
    }

    override fun deselectCurrentSlot() {
        hideTooltip()
        setWindowDeselected(container)
    }

    override fun selectCurrentSlot() {
        hasJustUpdated = true
    }

    override fun hideTooltip() {
        hasJustUpdated = false
        tooltip.hide()
    }

    fun update() {
        selectedHero = getSelectedHero()
        table.clear()
        fillRows()
    }

    protected abstract fun fillRows()

    fun addExtraToTable(totalExtra: Int) {
        if (totalExtra > 0) {
            val label = Label("+$totalExtra", LabelStyle(font, Color.FOREST))
            table.add(label).row()
        } else if (totalExtra < 0) {
            val label = Label(totalExtra.toString(), LabelStyle(font, Color.FIREBRICK))
            table.add(label).row()
        } else {
            table.add("").row()
        }
    }

    fun createImageOf(id: String): Image {
        val textureRegion = resourceManager.getAtlasTexture(id.lowercase())
        return Image(textureRegion).apply {
            setScaling(Scaling.none)
        }
    }

    private fun createSkin(): Skin {
        return Skin().apply {
            add("default", LabelStyle(font, Color.BLACK))
        }
    }

}
